cmd/binder/main.go

```diff
@@ -0,0 +1,290 @@
+func (b *Binding) BindInstanceMethod(f *File, d *ast.FuncDecl) {
+	r := List(Id("r"), Id("ok")).Op(":=").Add(Id("receiver")).Assert(Op("*").Id(b.ClassName)).Line()
+	r = r.If(Op("!").Id("ok")).Block(
+		Panic(
+			Qual("fmt", "Sprintf").Call(Lit("Impossible receiver type. Wanted "+b.ClassName+" got %s"), Id("receiver")),
+		),
+	).Line()
+	b.body(r, f, d)
+}
+
+func wrongArgNum(want int) Code {
+	return Return(Id("t").Dot("VM").Call().Dot("InitErrorObject").Call(
+		Qual(errorsPkg, "ArgumentError"),
+		Id("line"),
+		Qual(errorsPkg, "WrongNumberOfArgumentFormat"),
+		Lit(want),
+		Id("len").Call(Id("args")),
+	))
+}
+
+func wrongArgType(name, want string) Code {
+	return Return(Id("t").Dot("VM").Call().Dot("InitErrorObject").Call(
+		Qual(errorsPkg, "TypeError"),
+		Id("line"),
+		Qual(errorsPkg, "WrongArgumentTypeFormat"),
+		Lit(want),
+		Id(name).Dot("Class").Call().Dot("Name"),
+	))
+}
+
+// body is a helper function for generating the common body of a method
+func (b *Binding) body(receiver *Statement, f *File, d *ast.FuncDecl) {
+	s := f.Func().Id(b.bindingName(d))
+	s = s.Params(
+		Id("receiver").Qual(vmPkg, "Object"),
+		Id("line").Id("int"),
+		Id("t").Op("*").Qual(vmPkg, "Thread"),
+		Id("args").Index().Qual(vmPkg, "Object"),
+	).Qual(vmPkg, "Object")
+
+	var args []*Statement
+	for i, a := range allArgs(d.Type.Params) {
+		if i == 0 {
+			continue
+		}
+		i = i - 1
+		c := List(Id(fmt.Sprintf("arg%d", i)), Id("ok")).Op(":=").Id("args").Index(Lit(i)).Assert(Id(a.kind))
+		c = c.Line()
+		c = c.If(Op("!").Id("ok")).Block(
+			Panic(Lit(fmt.Sprintf("Argument %d must be %s", i, a.kind))),
+		).Line()
+		args = append(args, c)
+	}
+
+	inner := receiver.If(Len(Id("args")).Op("!=").Lit(d.Type.Params.NumFields() - 1)).Block(
+		wrongArgNum(d.Type.Params.NumFields() - 1),
+	).Line()
+	argNames := []Code{
+		Id("t"),
+	}
+	for i, a := range args {
+		inner = inner.Add(a).Line()
+		argNames = append(argNames, Id(fmt.Sprintf("arg%d", i)))
+	}
+
+	inner = inner.Return(Id("r").Dot(d.Name.Name).Call(argNames...))
+	s.Block(inner)
+}
+
+// mapping generates the "init" portion of the bindings.
+// This will call hooks in the vm package to load the class definition at runtime.
+func mapping(b *Binding, pkg string) Code {
+	fnName := func(s string) string {
+		x := camelcase.Split(s)
+		return strings.ToLower(strings.Join(x, "_"))
+	}
+
+	cm := Dict{}
+	for _, d := range b.ClassMethods {
+		cm[Lit(fnName(d.Name.Name))] = Id(b.bindingName(d))
+	}
+	im := Dict{}
+	for _, d := range b.InstanceMethods {
+		im[Lit(fnName(d.Name.Name))] = Id(b.bindingName(d))
+	}
+	dm := Qual(vmPkg, "RegisterExternalClass").Call(
+		Line().Lit(pkg),
+		Qual(vmPkg, "ExternalClass").Call(
+			Line().Lit(b.ClassName),
+			Line().Lit(pkg+".gb"),
+			Line().Map(String()).Qual(vmPkg, "Method").Values(cm),
+			Line().Map(String()).Qual(vmPkg, "Method").Values(im),
+		),
+	)
+	l := Func().Id("init").Params().Block(
+		dm,
+	)
+	return l
+}
+
+func main() {
```

Comment:
Can you add binder's usage document above?

Reply:
I mean like what's the scenario to use it, what do you need for using it and what options do you have..etc.